src/web100srv.c

```diff
@@ -1527,647 +1326,364 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
-        continue;
-      } else {
-        if (waiting > 0) goto ChldRdy;
-        continue;
-      }
+/** Goes from head to tail, reaping dead clients from the queue of clients and
+ * then updating all clients regarding their (possibly new) queue position. */
+void queue_maintenance(struct ndtchild **head, struct ndtchild **tail) {
+  ParentChildSignal child_message;
+  int queue_position;
+  int client_count = 0;
+  int running_count = 0;
+  int max_simultaneous_tests;
+  struct ndtchild *tmp_ptr, *prev_ptr;
+  // Walk the list of clients, reaping dead clients.
+  tmp_ptr = *head;
+  prev_ptr = NULL;
+  while (tmp_ptr != NULL) {
+    if (is_live_client(tmp_ptr)) {
+      prev_ptr = tmp_ptr;
+      tmp_ptr = tmp_ptr->next;
     } else {
-      log_println(3,
-                  "New connection received, waiting for accept() to complete");
-      if ((waiting > 0) && (testing == 0)) {
-        // no clients waiting, no test in progress
-        goto ChldRdy;
-      }
-      clilen = sizeof(cli_addr);
-      memset(&cli_addr, 0, clilen);
-      log_println(6, "Select() found %d clients ready, highest fd=%d", retcode,
-                  listenfd);
-      if (retcode > 1) {
-        for (i = 3; i <= listenfd; i++) {
-          if (FD_ISSET(i, &rfd)) {
-            listenfd = i;
-            break;
-          }
-        }
+      // Remove tmp_ptr from the linked list
+      if (prev_ptr == NULL) {
+        *head = (*head)->next;
+      } else {
+        prev_ptr->next = tmp_ptr->next;
       }
-      for (i = 0; i < RETRY_COUNT; i++) {
-        ctlsockfd = 0;
-        ctlsockfd = accept(listenfd, (struct sockaddr *)&cli_addr, &clilen);
-        // socket interrupted, retry
-        if ((ctlsockfd == -1) && (errno == EINTR)) continue;
-
-        // get meta test details copied into results
-        memcpy(&meta.c_addr, &cli_addr, clilen);
-        meta.family = ((struct sockaddr *)&cli_addr)->sa_family;
-
-        memset(rmt_addr, 0, sizeof(rmt_addr));
-        // get addr details based on socket info available
-        I2Addr tmp_addr = I2AddrBySockFD(get_errhandle(), ctlsockfd, False);
-        addr2a(&cli_addr, rmt_addr, sizeof(rmt_addr));
-
-        size_t rmt_host_strlen = sizeof(rmt_host);
-        memset(rmt_host, 0, rmt_host_strlen);
-        I2AddrNodeName(tmp_addr, rmt_host, &rmt_host_strlen);
-        /* I2AddrFree(tmp_addr); */
-        log_println(4, "New connection received from 0x%x [%s] sockfd=%d.",
-                    tmp_addr, tmpstr, ctlsockfd);
-        break;
+      if (tmp_ptr->next == NULL) {
+        *tail = prev_ptr;
       }
-
-      // verify that accept really worked and don't process connections
-      // that've failed
-      // RAC 2/11/10
-      if (ctlsockfd <= 0) {
-        log_println(4, "New connection request failed sockfd=%d reason-%d.",
-                    ctlsockfd, errno);
-        continue;
+      // Discard the client
+      close(tmp_ptr->pipe);
+      free(tmp_ptr);
```

Comment:
We're repeating this `ndtchild` cleanup code a lot. Can we have a `free_ndtchild` function that encapsulates this?